addons/point_of_sale/tests/test_pos_stock_account.py

```diff
@@ -235,3 +235,312 @@ def test_03_order_product_w_owner(self):
 
         # close the session
         self.pos_session.action_pos_session_validate()
+
+
+@odoo.tests.tagged('post_install', '-at_install')
+class TestPoSFailedStockPicking(TestPoSCommon):
+    """ Tests for anglo-saxon accounting when a stock picking fails when closing a POS session. These tests check that
+    the COGS lines are corrected when the user fixes the failed stock picking afterwards.
+    """
+    def setUp(self):
+        super(TestPoSFailedStockPicking, self).setUp()
+
+        # Using the 'closing' setting because the main use case where the correction is needed is when you have multiple
+        # orders in the session but one of them results in a failed stock picking, resulting in the inventory valuation
+        # for the entire session amounting to 0 when the 'closing' method is used. If the 'real' method is used, the
+        # impact is more limited.
+        self.company_data['company'].write({
+            'point_of_sale_update_stock_quantities': 'closing',
+        })
+
+        self.config = self.basic_config
+        # Product 1 requires a serial number for tracking purposes
+        self.product1 = self.create_product('Product 1', self.categ_anglo, 20.0, 10.0)
+        self.product1.tracking = 'serial'
+        # Product 2 doesn't have any tracking
+        self.product2 = self.create_product('Product 2', self.categ_anglo, 30.0, 15.0)
+        self.product2.tracking = 'none'
+        # start inventory with 1 item for each product
+        self.adjust_inventory([self.product1, self.product2], [1, 1])
+
+        self.output_account = self.categ_anglo.property_stock_account_output_categ_id
+        self.expense_account = self.categ_anglo.property_account_expense_categ_id
+        self.valuation_account = self.categ_anglo.property_stock_valuation_account_id
+
+    def test_selling_a_product_with_a_valid_picking(self):
+        self.open_new_session()
+
+        # create orders
+        orders = [self.create_ui_order_data([(self.product2, 1)])]
+
+        # sync orders
+        self.env['pos.order'].create_from_ui(orders)
+
+        # close the session
+        self.pos_session.action_pos_session_validate()
+
+        # picking should be in done state
+        self.assertEqual(self.pos_session.picking_ids[0].state, 'done', 'Picking should be in done state.')
+
+        self._assert_account_move(self.pos_session.move_id, {
+            'line_ids': [
+                {'account_id': self.sales_account.id, 'debit': 0, 'credit': 30, 'reconciled': False},
+                {'account_id': self.expense_account.id, 'debit': 15, 'credit': 0, 'reconciled': False},
+                {'account_id': self.cash_pm1.receivable_account_id.id, 'debit': 30, 'credit': 0, 'reconciled': True},
+                {'account_id': self.output_account.id, 'debit': 0, 'credit': 15, 'reconciled': True},
+            ],
+        })
```

Comment:
I don't know if PoS does it but we usually align stuff for easier comparison.
(at least, I did it in one of the test in their module and JCB agreed with it.)

Reply:
You mean align the values in columns inside the dict? I usually avoid that kind of formatting because it can be annoying to maintain and combine with code formatting tools. But if it's the convention for this kind of data I can update it, sure.

Reply:
Check what is done in other pos test file.
If both are done, do as you prefer.

Reply:
It's not aligned in the other tests in the same file: https://github.com/odoo/odoo/blob/e985004a0b13f55d2974936e21af3f46206e1bf2/addons/point_of_sale/tests/test_pos_stock_account.py#L180 , so I'll keep it consistent with that.